shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,313 @@
+     * that reaction when new tasks arrive until it is true, or a change elsewhere changes which Provider to run.
+     *
+     * @tparam T the condition expression that this reaction will make true before leaving.
+     */
+    template <typename State, enum State::Value value>
+    struct Causing {
+
+        /**
+         * Bind a causing expression in the Director.
+         *
+         * @tparam DSL the DSL from NUClear
+         *
+         * @param reaction the reaction that is having this when condition bound to it
+         */
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+            // Tell the director
+            reaction->reactor.powerplant.emit(
+                std::make_unique<commands::CausingExpression>(reaction, typeid(State), value));
+        }
+    };
+
+    /**
+     * A Task to be executed by the director, and a priority with which to execute this task.
+     *
+     * There are two different types of tasks that can be created using this emit, root level tasks and subtasks.
+     *
+     * Root level tasks:
+     * These are created when a reaction that is not a Provider (not a Provides, Entering or Leaving) emits the
+     * task. These tasks form the root of the execution tree and their needs will be met on a highest priority first
+     * basis. These tasks will persist until the Provider that they use emits a done task, or the task is re-emitted
+     * with a priority of 0.
+     *
+     * Subtasks:
+     * These are created when a Provider task emits a task to complete. These tasks must be emitted each time that
+     * reaction is run to persist as if that reaction runs again and does not emit these tasks they will be
+     * cancelled. Within these tasks the priority is used to break ties between two subtasks that share the same
+     * root task. In these cases the subtask that requested the Provider with the highest priority will have its
+     * task executed. The other subtask will be blocked until the active task is no longer in the call queue.
+     *
+     * If a subtask is emitted with optional then it is compared differently to other tasks when it comes to priority.
+     * This task and all its decendents will be considered optional. If it is compared to a task that does not have
+     * optional in it's parentage, the non optional task will win. However decendents of this task that are not optional
+     * will compare to eachother as normal. If two tasks both have optional in their parentage they will be compared as
+     * normal.
+     *
+     * @tparam T the Provider type that this task is for
+     */
+    template <typename T>
+    struct Task {
+
+        /**
+         * Emits a new task for the Director to handle.
+         *
+         * @param powerplant the powerplant object this emit is being called on
```

Comment:
We can have more than one `PowerPlant`? I thought this was a singleton object?

Reply:
It is, probably could change the comment here to describe it's the NUClear context you have in this function